### PR TITLE
fix(issue-1052): Single topic for RPC Message

### PR DESCRIPTION
**Pull request by AlejandroCabeza**
## Description
Rename RPC's Message `topicIds` to `topicId`.
Please, do look carefully into this one as I'm not sure how a well covered by tests is the repo, and my changes could be breaking stuff.

# Issue
Closes (hopefully): https://github.com/status-im/nim-libp2p/issues/1052

### CODE REVIEW

File: libp2p/protocols/pubsub/rpc/messages.nim [hunk 1]
```diff
@@ -27,52 +27,52 @@ proc expectedFields[T](t: typedesc[T], existingFieldNames: seq[string]) {.raises
...
+    subscribe*: bool
+    topic*: string
+
+  MessageId* = seq[byte]
+
+  Message* = object
+    fromPeer*: PeerId
+    data*: seq[byte]
+    seqno*: seq[byte]
+    topicId*: string
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/rpc/messages.nim:44]: good opportunity to line up the field name with the spec protobuf name

File: libp2p/protocols/pubsub/rpc/messages.nim [hunk 2]
```diff
@@ -113,7 +113,7 @@ func shortLog*(msg: Message): auto =
     fromPeer: msg.fromPeer.shortLog,
     data: msg.data.shortLog,
     seqno: msg.seqno.shortLog,
-    topicIds: $msg.topicIds,
+    topicId: $msg.topicId,
```
Comment (lchenut) [inline libp2p/protocols/pubsub/rpc/messages.nim:116]: I'm nitpicking, but `$` isn't necessary anymore, topicId is already a string.

File: libp2p/protocols/pubsub/floodsub.nim [hunk 3]
```diff
@@ -148,12 +148,13 @@ method rpcHandler*(f: FloodSub,
...
     var toSendPeers = initHashSet[PubSubPeer]()
-    for t in msg.topicIds:                     # for every topic in the message
-      if t notin f.topics:
-        continue
-      f.floodsub.withValue(t, peers): toSendPeers.incl(peers[])
+    let topic = msg.topicId
+    if topic notin f.topics:
+      continue
+    f.floodsub.withValue(topic, peers):
+      toSendPeers.incl(peers[])
```
Comment (lchenut) [inline libp2p/protocols/pubsub/floodsub.nim:155]: Hum... it looks wrong to me. The `continue` was for the loop you removed. Something like that should fix it.
```suggestion
    if topic in f.topics:
      f.floodsub.withValue(topic, peers):
        toSendPeers.incl(peers[])
      await handleData(f, topic, msg.data)
```

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 4]
```diff
@@ -344,18 +344,18 @@ proc validateAndRelay(g: GossipSub,
     # store in cache only after validation
     g.mcache.put(msgId, msg)
 
-    g.rewardDelivered(peer, msg.topicIds, true)
+    let topic = msg.topicId
+    g.rewardDelivered(peer, [topic], true)
```
Comment (lchenut) [inline libp2p/protocols/pubsub/gossipsub.nim:348]: It might be interesting to change `rewardDelivered` to only take one topics instead of an openarray. Since it's used only on this file.

File: libp2p/protocols/pubsub/floodsub.nim (hunk 3 above)
Comment (lchenut) [reply libp2p/protocols/pubsub/floodsub.nim:155]: Nevermind, `f.broadcast` will do nothing if `topic` isn't in `f.topics`

File: libp2p/protocols/pubsub/rpc/messages.nim (hunk 2 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/rpc/messages.nim:116]: Do nitpick, that's what we're here for :)

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 4 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub.nim:348]: I did think on changing this, but for some reason didn't check whether it was called somewhere else and then I forgot 🤦‍♀️

If you guys think this is not going to be called with multiple topics then I'm all for updating it. Do confirm, please :)

Comment (arnetheduck) [reply libp2p/protocols/pubsub/gossipsub.nim:348]: +1 to change

**Comment by arnetheduck** [comment]
one thing missing here is that the code must check that there is (at least) one topic when decoding the message from bytes

File: libp2p/protocols/pubsub/gossipsub/scoring.nim [hunk 5]
```diff
@@ -21,16 +21,56 @@ import ../pubsub
...
-declareCounter(libp2p_gossipsub_bad_score_disconnection, "the number of peers disconnected by gossipsub", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_firstMessageDeliveries, "Detailed gossipsub scoring metric", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_meshMessageDeliveries, "Detailed gossipsub scoring metric", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_meshFailurePenalty, "Detailed gossipsub scoring metric", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_invalidMessageDeliveries, "Detailed gossipsub scoring metric", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_appScore, "Detailed gossipsub scoring metric", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_behaviourPenalty, "Detailed gossipsub scoring metric", labels = ["agent"])
-declareGauge(libp2p_gossipsub_peers_score_colocationFactor, "Detailed gossipsub scoring metric", labels = ["agent"])
-declarePublicCounter(libp2p_gossipsub_peers_rate_limit_hits, "The number of times peers were above their rate limit", labels = ["agent"])
+declareGauge(
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub/scoring.nim:24]: these (nph-like) changes should probably not be part of this PR

File: libp2p/protocols/pubsub/rpc/messages.nim (hunk 1 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/rpc/messages.nim:44]: On the one hand, we can't change `fromPeer` to `from` as the second one is a reserved keyword in nim.

On the other, `go-lib-p2p` does use `Message.topic` ([ref](https://github.com/libp2p/go-libp2p-pubsub/blob/048a4d30d0c3c00829181fd81aa697eb60497448/pb/rpc.proto#L21)), but uses `ControlIHave.topicID` ([ref](https://github.com/libp2p/go-libp2p-pubsub/blob/048a4d30d0c3c00829181fd81aa697eb60497448/pb/rpc.proto#L34)). I don't know which one I like more, but I do know I'd stick to using one and the same for both cases (as we're already doing).

File: libp2p/protocols/pubsub/gossipsub/scoring.nim (hunk 5 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:24]: Yeah, my bad, styling tool in vscode running when it pleases 🤦‍♀️ Reverting

File: libp2p/protocols/pubsub/rpc/messages.nim (hunk 1 above)
Comment (arnetheduck) [reply libp2p/protocols/pubsub/rpc/messages.nim:44]: The canonical names for each field are in the spec: 

https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-message

It doesn't greatly matter what go-libp2p does, it's not authoritative. 

Technically, we can use \`from\` (in backticks), though it's a bit of a bother in nim, so for that field in particular, we have reason to not go with the canonical name).

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/rpc/messages.nim:44]: I did check the spec, but also double checked what other libraries were doing given the `from` situation (given I wasn't aware of the backtick trick).

I'm fing changing it to `topic` tbh, doing it.

Comment (arnetheduck) [reply libp2p/protocols/pubsub/rpc/messages.nim:44]: We will likely move to a model where we generate fields from the `.proto` directly so this is a step in the right direction. It's also useful when using the [protobuf JSON mapping](https://protobuf.dev/programming-guides/proto3/#json) - again, not for today, but good habit in general.

File: libp2p/protocols/pubsub/rpc/messages.nim [hunk 6]
```diff
@@ -27,52 +27,52 @@ proc expectedFields[T](t: typedesc[T], existingFieldNames: seq[string]) {.raises
...
+
+  ControlMessage* = object
+    ihave*: seq[ControlIHave]
+    iwant*: seq[ControlIWant]
+    graft*: seq[ControlGraft]
+    prune*: seq[ControlPrune]
+    idontwant*: seq[ControlIWant]
+
+  ControlIHave* = object
+    topic*: string
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/rpc/messages.nim:56]: The gossipsub fields are named different from floodsub (https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md#protobuf)

**Comment by arnetheduck** [comment]
> What's the expected behaviour if there isn't?

penalize the peer for sending an invalid message (same as if signature was bad or it wasn't protobuf at all)

File: libp2p/protocols/pubsub/rpc/messages.nim (hunk 6 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/rpc/messages.nim:56]: Oh, alright. That I did not check, thanks for the heads up.

File: libp2p/protocols/pubsub/gossipsub/scoring.nim [hunk 7]
```diff
@@ -250,43 +250,46 @@ proc punishInvalidMessage*(g: GossipSub, peer: PubSubPeer, msg: Message) {.async
...
+    peer: PubSubPeer,
+    topic: string,
+    first: bool,
+    delay = ZeroDuration,
+) =
+  let t = topic
+  if t notin g.topics:
+    return
+
+  let topic = t
```
Comment (lchenut) [inline libp2p/protocols/pubsub/gossipsub/scoring.nim:276]: These two assignations `let t = topic` and then `let topic = t` confuse me so much, do you have any idea why it was this way?

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:276]: No, and when I simplified it it stopped working 😂😂

File: libp2p/protocols/pubsub/rpc/protobuf.nim [hunk 8]
```diff
@@ -288,7 +288,8 @@ proc decodeMessage*(pb: ProtoBuffer): ProtoResult[Message] {.inline.} =
   if ?pb.getField(4, msg.topic):
     trace "decodeMessage: read topic", topic = msg.topic
   else:
-    trace "decodeMessage: topic is missing"
+    error "decodeMessage: topic is required but missing"
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/rpc/protobuf.nim:291]: broadly we don't log errors for something that is under the control of a remote peer as this would allow spamming users (who are normally assumed to run on `INF` level) - trace is the right level here (or possibly debug, since it should be uncommon)

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/rpc/protobuf.nim:291]: Oh, alright, will get it updated.

File: libp2p/protocols/pubsub/gossipsub/scoring.nim (hunk 7 above)
Comment (lchenut) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:276]: So I dug a bit into it and there's two things related:
- `let t = topic` : https://github.com/vacp2p/nim-libp2p/issues/581
- `let topic = t` : https://github.com/vacp2p/nim-libp2p/commit/ce42674d80222e5dc6f9c779cef128005d696bf9
Both were issues raised with nim1.4, I can understand why the first hasn't been fixed (or even if it's fixable), but the second seems a bit odd. What happens if you just remove the second assignation?

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim]: I ended up removing the two "unnecessary" assigments and tests seemed to pass locally. I did that once previously and tests failed, so... I don't know. Let's see what the CI says.

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim]: Passed, amazing. All good then.

File: libp2p/protocols/pubsub/gossipsub/scoring.nim [hunk 9]
```diff
@@ -250,43 +250,43 @@ proc punishInvalidMessage*(g: GossipSub, peer: PubSubPeer, msg: Message) {.async
         await g.disconnectPeer(peer)
         raise newException(PeerRateLimitError, "Peer disconnected because it's above rate limit.")
 
+  let topic = msg.topic
+  # TODO: REVIEW
+  if topic notin g.topics:
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub/scoring.nim:255]: even if the topic is not in topics, `invalidMessageDeliveries` needs to be incremented

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 10]
```diff
@@ -170,7 +200,9 @@ method onNewPeer*(g: GossipSub, peer: PubSubPeer) =
   peer.iHaveBudget = IHavePeerBudget
   peer.pingBudget = PingsPeerBudget
 
-method onPubSubPeerEvent*(p: GossipSub, peer: PubSubPeer, event: PubSubPeerEvent) {.gcsafe.} =
+method onPubSubPeerEvent*(
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:203]: these are unrelated reformattings that should not be part of this PR

Comment (diegomrsantos) [reply libp2p/protocols/pubsub/gossipsub.nim:203]: @AlejandroCabeza could you please revert all formatting changes?

File: libp2p/protocols/pubsub/floodsub.nim (hunk 3 above)
Comment (diegomrsantos) [reply libp2p/protocols/pubsub/floodsub.nim:155]: Could you please share why `f.broadcast` won't do anything? Even if it's true, IMO we shouldn't rely on that and change the previous behavior without a good reason for doing so.

Comment (lchenut) [reply libp2p/protocols/pubsub/floodsub.nim:155]: Because `broadcast` [will update some metrics](https://github.com/vacp2p/nim-libp2p/blob/unstable/libp2p/protocols/pubsub/pubsub.nim#L170-L210) by incrementing by the number of `toSendPeers` which is zero.
Then it iterates through `toSendPeers` [to send data](https://github.com/vacp2p/nim-libp2p/blob/unstable/libp2p/protocols/pubsub/pubsub.nim#L212-L219) (encoded or not) but it is empty so it won't do anything.

Comment (diegomrsantos) [reply libp2p/protocols/pubsub/floodsub.nim:155]: Thanks, you are right. But it'd still log. Maybe adding a log when `topic` isn't in `f.topics` is enough.

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/floodsub.nim:155]: Added

File: libp2p/protocols/pubsub/gossipsub/scoring.nim (hunk 9 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:255]: But that was not what was happening before:
```nim
for tt in msg.topicIds:
    let t = tt
    if t notin g.topics:
      continue

    let tt = t
    # update stats
    g.withPeerStats(peer.peerId) do (stats: var PeerStats):
      stats.topicInfos.mgetOrPut(tt, TopicInfo()).invalidMessageDeliveries += 1
```
Only topics in `g.topics` are counted in `invalidMessageDeliveries`. 
Was the previous code wrong?

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 10 above)
Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub.nim:203]: Sure, I'll try to undo the stylings.

Comment (diegomrsantos) [reply libp2p/protocols/pubsub/gossipsub.nim:203]: I think the formatting when the code also changed is fine, I, at least, I was referring to the changes that are purely formatting related.

File: libp2p/protocols/pubsub/gossipsub/scoring.nim (hunk 9 above)
Comment (arnetheduck) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:255]: you're right - I misread the indent level

Comment (arnetheduck) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:255]: let's get rid of this TODO then

Comment (AlejandroCabeza) [reply libp2p/protocols/pubsub/gossipsub/scoring.nim:254]: Done.

**Review by lchenut — APPROVED** [review]
LGTM

**Review by arnetheduck — APPROVED** [review]
